### PR TITLE
Add table aws_drs_source_server . closes #1457

### LINKED ISSUE #1457: Add table aws_drs_source_server
**References**

https://github.com/aws/aws-sdk-go-v2/blob/main/service/drs/api_op_DescribeSourceServers.go#L16

### PR DESCRIPTION
**Pull request by sdsubhajitdas**
# Integration test logs
<details>
  <summary>Logs</summary>

```
Add passing integration test logs here
```
</details>

# Example query results
<details>
  <summary>Results</summary>

```
> select
  arn,
  last_launch_result,
  source_server_id,
  title
from
  aws_drs_source_server;
+----------------------------------------------------------------------+--------------------+---------------------+---------------------+
| arn                                                                  | last_launch_result | source_server_id    | title               |
+----------------------------------------------------------------------+--------------------+---------------------+---------------------+
| arn:aws:drs:us-east-1:123456789012:source-server/s-453a73d0aa5143bc0 | "NOT_STARTED"      | s-453a73d0aa5143bc0 | s-453a73d0aa5143bc0 |
+----------------------------------------------------------------------+--------------------+---------------------+---------------------+
> select
  arn,
  title,
  source_cloud_properties ->> 'OriginAccountID' as source_cloud_properties_origin_account_id,
  source_cloud_properties ->> 'OriginAvailabilityZone' as source_cloud_properties_origin_availability_zone,
  source_cloud_properties ->> 'OriginRegion' as source_cloud_properties_origin_region
from
  aws_drs_source_server;
+----------------------------------------------------------------------+---------------------+-------------------------------------------+--------------------------------------------------+---------------------------------------+
| arn                                                                  | title               | source_cloud_properties_origin_account_id | source_cloud_properties_origin_availability_zone | source_cloud_properties_origin_region |
+----------------------------------------------------------------------+---------------------+-------------------------------------------+--------------------------------------------------+---------------------------------------+
| arn:aws:drs:us-east-1:123456789012:source-server/s-453a73d0aa5143bc0 | s-453a73d0aa5143bc0 | 123456789012                              | us-east-1a                                       | us-east-1                             |
+----------------------------------------------------------------------+---------------------+-------------------------------------------+--------------------------------------------------+---------------------------------------+
> select
  arn,
  title,
  source_properties ->> 'Cpus' as source_cpus,
  source_properties ->> 'Disks' as source_disks,
  source_properties -> 'IdentificationHints' ->> 'Hostname' as source_hostname,
  source_properties ->> 'NetworkInterfaces' as source_network_interfaces,
  source_properties -> 'Os' ->> 'FullString' as source_os,
  source_properties -> 'RamBytes' as source_ram_bytes,
  source_properties -> 'RecommendedInstanceType' as source_recommended_instance_type,
  source_properties -> 'LastUpdatedDateTime' as source_last_updated_date_time
from
  aws_drs_source_server;
+----------------------------------------------------------------------+---------------------+--------------------------------------------------------------------------+----------------------------------------------------+----------------------------+---------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------+----------------------------------+------------------------------------+
| arn                                                                  | title               | source_cpus                                                              | source_disks                                       | source_hostname            | source_network_interfaces                                                       | source_os                                                                                                                                                                                         | source_ram_bytes | source_recommended_instance_type | source_last_updated_date_time      |
+----------------------------------------------------------------------+---------------------+--------------------------------------------------------------------------+----------------------------------------------------+----------------------------+---------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------+----------------------------------+------------------------------------+
| arn:aws:drs:us-east-1:123456789012:source-server/s-453a73d0aa5143bc0 | s-453a73d0aa5143bc0 | [{"Cores": 1, "ModelName": "Intel(R) Xeon(R) CPU E5-2676 v3 @ 2.40GHz"}] | [{"Bytes": 8589934592, "DeviceName": "/dev/xvda"}] | ip-10-0-15-53.ec2.internal | [{"Ips": ["10.0.15.53"], "IsPrimary": true, "MacAddress": "02-FD-0C-48-D2-B5"}] | Linux version 5.10.147-133.644.amzn2.x86_64 (mockbuild@ip-10-0-33-205) (gcc10-gcc (GCC) 10.3.1 20210422 (Red Hat 10.3.1-1), GNU ld version 2.35-21.amzn2.0.1) #1 SMP Fri Oct 14 01:16:24 UTC 2022 | 1012649984       | "c5.large"                       | "2022-12-02T15:00:10.714955+00:00" |
+----------------------------------------------------------------------+---------------------+--------------------------------------------------------------------------+----------------------------------------------------+----------------------------+---------------------------------------------------------------------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+------------------+----------------------------------+------------------------------------+
> select
  arn,
  title,
  data_replication_info -> 'DataReplicationInitiation' ->> 'StartDateTime' as data_replication_start_date_time,
  data_replication_info -> 'DataReplicationInitiation' ->> 'NextAttemptDateTime' as data_replication_next_attempt_date_time,
  data_replication_info ->> 'DataReplicationError' as data_replication_error,
  data_replication_info ->> 'DataReplicationState' as data_replication_state,
  data_replication_info ->> 'ReplicatedDisks' as data_replication_replicated_disks
from
  aws_drs_source_server;
+----------------------------------------------------------------------+---------------------+----------------------------------+-----------------------------------------+------------------------+------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| arn                                                                  | title               | data_replication_start_date_time | data_replication_next_attempt_date_time | data_replication_error | data_replication_state | data_replication_replicated_disks                                                                                                                                      |
+----------------------------------------------------------------------+---------------------+----------------------------------+-----------------------------------------+------------------------+------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| arn:aws:drs:us-east-1:123456789012:source-server/s-453a73d0aa5143bc0 | s-453a73d0aa5143bc0 | 2022-12-01T09:59:09.475408+00:00 | 2022-12-02T15:52:00+00:00               | <null>                 | CONTINUOUS             | [{"DeviceName": "/dev/xvda", "TotalStorageBytes": 8589934592, "RescannedStorageBytes": 8589934592, "BackloggedStorageBytes": 0, "ReplicatedStorageBytes": 8589934592}] |
+----------------------------------------------------------------------+---------------------+----------------------------------+-----------------------------------------+------------------------+------------------------+------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
> select
  title,
  arn,
  last_launch_result,
  source_server_id
from
  aws_drs_source_server
where
  last_launch_result = 'FAILED';
+---------------------+----------------------------------------------------------------------+--------------------+---------------------+
| title               | arn                                                                  | last_launch_result | source_server_id    |
+---------------------+----------------------------------------------------------------------+--------------------+---------------------+
| s-3a960fb732dc6d6ed | arn:aws:drs:us-east-1:123456789012:source-server/s-3a960fb732dc6d6ed | FAILED             | s-3a960fb732dc6d6ed |
+---------------------+----------------------------------------------------------------------+--------------------+---------------------+
> select
  title,
  arn,
  data_replication_info ->> 'DataReplicationState' as data_replication_state,
  data_replication_info ->> 'DataReplicationError' as data_replication_error,
  data_replication_info -> 'DataReplicationInitiation' ->> 'StartDateTime' as data_replication_start_date_time,
  data_replication_info -> 'DataReplicationInitiation' ->> 'NextAttemptDateTime' as data_replication_next_attempt_date_time
from
  aws_drs_source_server
where
  data_replication_info ->> 'DataReplicationState' = 'DISCONNECTED';
+---------------------+----------------------------------------------------------------------+------------------------+------------------------+----------------------------------+-----------------------------------------+
| title               | arn                                                                  | data_replication_state | data_replication_error | data_replication_start_date_time | data_replication_next_attempt_date_time |
+---------------------+----------------------------------------------------------------------+------------------------+------------------------+----------------------------------+-----------------------------------------+
| s-3a960fb732dc6d6ed | arn:aws:drs:us-east-1:123456789012:source-server/s-3a960fb732dc6d6ed | DISCONNECTED           | <null>                 | 2022-12-08T11:33:20.400791+00:00 | 2022-12-08T11:34:00+00:00               |
+---------------------+----------------------------------------------------------------------+------------------------+------------------------+----------------------------------+-----------------------------------------+


```
</details>

### CODE REVIEW

File: aws/table_aws_drs_source_server.go [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+			{
+				Name:        "recovery_instance_id",
+				Description: "The ID of the Recovery Instance associated with this Source Server.",
+				Type:        proto.ColumnType_STRING,
+				Transform:   transform.FromField("RecoveryInstanceId"),
+			},
+			{
+				Name:        "source_properties",
+				Description: "The source properties of the Source Server.",
+				Type:        proto.ColumnType_JSON,
```
Comment (cbruno10) [inline aws/table_aws_drs_source_server.go:49]: This table has a lot of JSON columns, so when writing example docs, please be sure to create examples showcasing how to dig into key JSON columns, e.g., https://hub.steampipe.io/plugins/turbot/mongodbatlas/tables/mongodbatlas_cluster#get-auto-scaling-details-of-all-clusters

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@sdsubhajitdas I saw the PR was still in draft, but wanted to leave some early suggestions in case they're helpful, thanks!

File: aws/table_aws_drs_source_server.go [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+	"github.com/turbot/steampipe-plugin-sdk/v4/plugin/transform"
+)
+
+func tableAwsDRSSourceServer(_ context.Context) *plugin.Table {
+	return &plugin.Table{
+		Name:        "aws_drs_source_server",
+		Description: "AWS DRS Source Server",
+		List: &plugin.ListConfig{
+			KeyColumns: []*plugin.KeyColumn{
+				{Name: "source_server_id", Require: plugin.Optional},
```
[the remark this answers is not in the document]
Comment (sdsubhajitdas) [reply aws/table_aws_drs_source_server.go:20]: I have added two new columns for this API

File: aws/table_aws_drs_source_server.go [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+			if limit < 1 {
+				maxItems = int32(1)
+			} else {
+				maxItems = int32(limit)
+			}
+		}
+	}
+
+	input.MaxResults = int32(maxItems)
+	sourceServerId := d.KeyColumnQualString("source_server_id")
```
Comment (cbruno10) [inline aws/table_aws_drs_source_server.go:156]:
```suggestion
	sourceServerID := d.KeyColumnQualString("source_server_id")
```

File: aws/table_aws_drs_source_server.go [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+	stagingAccountId := d.KeyColumnQualString("staging_account_id")
```
Comment (cbruno10) [inline aws/table_aws_drs_source_server.go:157]:
```suggestion
	stagingAccountID := d.KeyColumnQualString("staging_account_id")
```

File: aws/table_aws_drs_source_server.go [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+	hardwareId := d.KeyColumnQualString("hardware_id")
```
Comment (cbruno10) [inline aws/table_aws_drs_source_server.go:158]:
```suggestion
	hardwareID := d.KeyColumnQualString("hardware_id")
```

File: aws/table_aws_drs_source_server.go [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+		plugin.Logger(ctx).Error("aws_drs_source_server.listAwsDRSSourceServers", "connection_error", err)
+		return nil, err
+	}
+
+	if svc == nil {
+		// Unsupported region, return no data
+		return nil, nil
+	}
+
+	maxItems := int32(10000)
```
Comment (cbruno10) [inline aws/table_aws_drs_source_server.go:140]: Is 10,000 the max limit for this? The API doc doesn't seem to mention a limit, so can it go beyond 10,000?

File: docs/tables/aws_drs_source_server.md [hunk 7]
```diff
@@ -0,0 +1,63 @@
+# Table: aws_drs_source_server
+
+AWS Elastic Disaster Recovery (AWS DRS) minimizes downtime and data loss with fast, reliable recovery of on-premises and cloud-based applications using affordable storage, minimal compute, and point-in-time recovery.
+
+## Examples
```
Comment (cbruno10) [inline docs/tables/aws_drs_source_server.md:5]: Would adding an example using a `where` clause with the `last_launch_result` column be helpful, e.g., `### List source servers that failed last recovery launch`? Are there any other columns we could use in additional examples to show the health/status or security information of a source server?

**Review by cbruno10 — CHANGES_REQUESTED** [review]
@Subhajit97 Please see comments, thanks!

File: docs/tables/aws_drs_source_server.md (hunk 7 above)
Comment (sdsubhajitdas) [reply docs/tables/aws_drs_source_server.md:5]: @cbruno10 I have added two examples using the where clause.
- Using `last_launch_result ` field as you suggested.
- Using the `data_replication_state` field will help identify which source servers are disconnected from the data replication process.

**Review by cbruno10 — APPROVED** [review]

**Pull request merged**